src/viadot/sources/smb.py

```diff
@@ -0,0 +1,330 @@
+            None: If `date_filter` is None.
+
+        Raises:
+            ValueError: If `date_filter` is neither a string nor a tuple of exactly
+                two strings.
+        """
+        if date_filter is None:
+            return None
+
+        ddh = DynamicDateHandler(
+            dynamic_date_symbols=dynamic_date_symbols,
+            dynamic_date_format=dynamic_date_format,
+            dynamic_date_timezone=dynamic_date_timezone,
+        )
+
+        match date_filter:
+            case str():
+                return pendulum.parse(ddh.process_dates(date_filter)).date()
+
+            case (start, end) if isinstance(start, str) and isinstance(end, str):
+                return (
+                    pendulum.parse(ddh.process_dates(start)).date(),
+                    pendulum.parse(ddh.process_dates(end)).date(),
+                )
+
+            case _:
+                msg = (
+                    "date_filter must be a string, a tuple of exactly 2 dates, or None."
+                )
+                raise ValueError(msg)
+
+    def _scan_directory(
+        self,
+        path: str,
+        keywords: list[str] | None,
+        extensions: list[str] | None,
+        date_filter_parsed: pendulum.Date
+        | tuple[pendulum.Date, pendulum.Date]
+        | None = None,
+    ) -> None:
+        """Recursively scans a directory for matching files based on filters.
+
+        It applies keyword and extension filters and can filter files based on
+        modification dates.
+
+        Args:
+            path (str): The directory path to scan.
+            keywords (list[str] | None): List of keywords to search for in filenames.
+                Defaults to None.
+            extensions (list[str] | None): List of file extensions to filter by.
+                Defaults to None.
+            date_filter_parsed (
+                pendulum.Date | tuple[pendulum.Date, pendulum.Date] | None
+            ):
+                - A single `pendulum.Date` for exact date filtering.
+                - A tuple of two `pendulum.Date` values for date range filtering.
+                - None, if no date filter is applied.
+                Defaults to None.
+        """
+        try:
+            entries = self._get_directory_entries(path)
+            for entry in entries:
+                if entry.is_file() and self._is_matching_file(
+                    entry, keywords, extensions, date_filter_parsed
+                ):
+                    self._handle_matching_file(entry)
+                elif entry.is_dir():
+                    self._scan_directory(
+                        entry.path, keywords, extensions, date_filter_parsed
+                    )
+        except Exception as e:
+            self.logger.exception(f"Error scanning or downloading from {path}: {e}")  # noqa: TRY401
+
+    def _handle_matching_file(self, entry: smbclient._os.SMBDirEntry) -> None:
```

Comment:
This should be a pure function, so return the found files rather than storing them inside an attribute. Pure functions are easier to understand and test.

Reply:
[0a64549](https://github.com/dyvenia/viadot/pull/1127/commits/0a64549bafc4c3b560f9099b7059f700c63bad1f)

Reply:
- you still use `self.found_files.update(found_file)`
- I would use precise typing, so `dict` -> `dict[str, bytes]`
- TBH I think getting rid of `_fetch_file_content()` would simplify things, this func is only doing `open()`, you could just include this in `_get_file_content()`?

Reply:
`self.found_files` is removed here, after the last change: [b25fa32](https://github.com/dyvenia/viadot/pull/1127/commits/b25fa325c2409db2c4daec801eb02fbb07de221d)

You're right. It's used only for getting the content so it makes more sense to move it there.